### PR TITLE
enhancement(geoip enrichment): Add support for Connection-Type databases

### LINKED ISSUE #13702: geoip: GeoIP2-Connection-Type.mmdb doesn't work with geoip transform
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Using a `Connection-Type` database with the `geoip` transform results in an empty object with the same structure as a `City` database:

```json
{
  "asn": {
    "autonomous_system_number": 207586,
    "autonomous_system_organization": "Doyle (North West) Ltd.",
    "isp": "",
    "organization": ""
  },
  "connection_type": {
    "city_name": "",
    "continent_code": "",
    "country_code": "",
    "country_name": "",
    "latitude": "",
    "longitude": "",
    "metro_code": "",
    "postal_code": "",
    "region_code": "",
    "region_name": "",
    "timezone": ""
  },
  "geoip": {
    "city_name": "",
    "continent_code": "EU",
    "country_code": "GB",
    "country_name": "United Kingdom",
    "latitude": "51.4964",
    "longitude": "-0.1224",
    "metro_code": "",
    "postal_code": "",
    "region_code": "",
    "region_name": "",
    "timezone": "Europe/London"
  }
}
```

### Configuration

```text
[transforms.geoip]
database = "/nix/store/zhrqk5bfpwsxshmyvsz8w7dzy0f23mkv-GeoIP2-City.mmdb"
inputs = ["parse_log"]
source = "source_address"
target = "geoip"
type = "geoip"

[transforms.asn]
database = "/nix/store/27ilhy1skdl9g5piihyndpk0yyi4azpm-GeoLite2-ASN.mmdb"
inputs = ["geoip"]
source = "source_address"
target = "asn"
type = "geoip"

[transforms.connection_type]
database = "/nix/store/sagj9srjplk7rgkfx486sf9a394mlmyl-GeoIP2-Connection-Type.mmdb"
inputs = ["asn"]
source = "source_address"
target = "connection_type"
type = "geoip"
```

### Version

vector 0.23.0 (x86_64-unknown-linux-gnu)

### Debug Output

```text
2022-07-25T11:39:41.091116Z TRACE transform{component_kind="transform" component_id=connection_type component_type=geoip component_name=connection_type}: vector_common::internal_event::events_received: Events received. count=1 byte_size=4946
2022-07-25T11:39:41.091151Z TRACE transform{component_kind="transform" component_id=connection_type component_type=geoip component_name=connection_type}: vector_buffers::topology::channel::limited_queue: Sent item.
2022-07-25T11:39:41.091162Z TRACE transform{component_kind="transform" component_id=connection_type component_type=geoip component_name=connection_type}: vector_core::fanout: Sent item to fanout.
2022-07-25T11:39:41.091172Z TRACE transform{component_kind="transform" component_id=connection_type component_type=geoip component_name=connection_type}: vector_common::internal_event::events_sent: Events sent. count=1 byte_size=5892 output=_default
```

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by w4**
Closes #13702

### CODE REVIEW

File: src/transforms/geoip.rs [hunk 1]
```diff
@@ -165,113 +188,141 @@ struct City<'a> {
...
+                emit!(ParserMissingFieldError {
+                    field: &self.source
+                });
 
-        let json_value = if self.has_isp_db() {
-            serde_json::to_value(isp)
-        } else {
-            serde_json::to_value(city)
+                None
+            }
```
Comment (tobz) [inline src/transforms/geoip.rs:322]: Hmmm, I think this is actually a breaking change, as all of the fields for the given MaxMind database (i.e. the city-specific fields if using the City database, etc) would still be added to the target field, albeit empty, if the IP wasn't parseable or if the source field was missing.

Now the target field will just be `null`.

@jszwedko What do you think? I don't have a strong opinion on `null` overall vs having all the target fields being present but empty.

**Comment by jszwedko** [comment]
Hey @w4 ! Apologies for not getting to this sooner. This `geoip` transform is actually deprecated now that we've we added support for enriching geoip data via enrichment tables (see: https://github.com/vectordotdev/vector/pull/13338). Would you mind moving these changes over there?

**Comment by w4** [comment]
Ah interesting, will do thanks

**Comment by w4** [comment]
This is updated now, thanks @jszwedko

**Review by jszwedko — APPROVED** [review]
This looks good, thanks for porting over the logic @w4 !

**Review by jszwedko — CHANGES_REQUESTED** [review]
Actually, I realized we need to update the docs here too: https://github.com/vectordotdev/vector/blob/e1268b1ad376669dc8069d0bf59d34dedc415b2e/website/cue/reference/remap/functions.cue#L60

I'll push an update.

**Review by jszwedko — APPROVED** [review]

**Comment by jszwedko** [comment]
@w4 I wasn't able to push to this PR branch for some reason. Would you mind applying the following diff:

```diff
0f3b36b3a - Document support for Connection-Type databases (5 minutes ago) <Jesse Szwedko>
diff --git a/website/cue/reference/remap/functions.cue b/website/cue/reference/remap/functions.cue
index d491c9873..f8fc7df8b 100644
--- a/website/cue/reference/remap/functions.cue
+++ b/website/cue/reference/remap/functions.cue
@@ -82,26 +82,27 @@ remap: {
 		and without a value, `null` will be used. This table can return following fields:
 
 		* ISP databases:
-
 			* `autonomous_system_number`
 			* `autonomous_system_organization`
 			* `isp`
 			* `organization`
 
 		* City databases:
-
 			* `city_name`
 			* `continent_code`
 			* `country_code`
 			* `country_name`
 			* `region_code`
 			* `region_name`
-			* `metro_code`		
+			* `metro_code`
 			* `latitude`
-			* `longitude`		
+			* `longitude`
 			* `postal_code`
 			* `timezone`
 
+		* Connection-Type databases:
+			* `connection_type`
+
 		To use this function, you need to update your Vector configuration to
 		include an
 		[`enrichment_tables`](\(urls.vector_configuration_global)/#enrichment_tables)
```

**Comment by w4** [comment]
Not a problem, that's all done - thanks @jszwedko

**Comment by jszwedko** [comment]
Great, thanks! I marked this to merge once CI passes.

**Comment by jszwedko** [comment]
Merging since CI failure was from master.

**Pull request merged**